metrics-sampler/src/onboarding_metrics_sampler.rs

```diff
@@ -0,0 +1,131 @@
+const NUM_WALLETS_METRIC: &str = "num_wallets";
+/// Metric describing the number of wallets with at least one deposit
+const NUM_WALLETS_WITH_DEPOSITS_METRIC: &str = "num_wallets_with_deposits";
+/// Metric describing the number of wallets with at least one order placement
+const NUM_WALLETS_WITH_ORDERS_METRIC: &str = "num_wallets_with_orders";
+/// Metric describing the number of wallets with at least one match
+const NUM_WALLETS_WITH_MATCHES_METRIC: &str = "num_wallets_with_matches";
+
+/// Samples onboarding metrics at a fixed interval
+#[derive(Clone)]
+pub struct OnboardingMetricsSampler {
+    /// A handle to the global state
+    state: State,
+}
+
+impl OnboardingMetricsSampler {
+    /// Create a new `OnboardingMetricsSampler`
+    pub fn new(state: State) -> Self {
+        Self { state }
+    }
+}
+
+impl AsyncMetricSampler for OnboardingMetricsSampler {
+    fn name(&self) -> &str {
+        ONBOARDING_METRICS_SAMPLER_NAME
+    }
+
+    fn interval(&self) -> Duration {
+        Duration::from_millis(ONBOARDING_METRICS_SAMPLE_INTERVAL_MS)
+    }
+
+    async fn sample(&self) -> Result<(), String> {
+        // First metric: number of wallets
+        // Second metric: number of wallets w/ at least one deposit
+        // Third metric: number of wallets w/ at least one order placement
+        // Fourth metric: number of wallets w/ at least one match
+        // Latter 3 metrics need task history. So we fetch all wallet IDs, task history
+        // for all wallets, and bucket into sets from there
+
+        let wallets = self.state.get_all_wallets().await?;
+        let num_wallets = wallets.len();
+
+        let mut num_wallets_with_deposits = 0;
+        let mut num_wallets_with_orders = 0;
+        let mut num_wallets_with_matches = 0;
+
+        for wallet in wallets {
+            // Short-circuit state queries for task history by checking if the wallet has
+            // outstanding fees. This implies that the wallet has experienced at
+            // least one match, which itself implies that that the wallet has
+            // had deposits and orders.
+            if wallet.has_outstanding_fees() {
+                num_wallets_with_deposits += 1;
+                num_wallets_with_orders += 1;
+                num_wallets_with_matches += 1;
+            } else {
+                // If the wallet does not have outstanding fees, we must query
+                // its task history to see if it has deposited /
+                // placed an order / matched.
+
+                // We fetch only the last 50 tasks for the wallet. This should be
+                // a sufficient sample to determine the wallet's onboarding state.
+                let tasks = self.state.get_task_history(50, &wallet.wallet_id).await?;
+
+                let mut has_deposited = false;
+                let mut has_placed_order = false;
+                let mut has_matched = false;
+
+                for task in tasks {
```

Comment:
similarly, we should break out this into a helper, `process_task_metrics` or similar. not a fan of long methods with lots of nested control flow